### PR TITLE
Introduce ObjectStoreProvider to create an object store based on the url

### LINKED ISSUE #2905: Make ObjectStoreRegistry as a trait which can allow Ballista to introduce a self registry ObjectStoreRegistry
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently ObjectStore cannot be self-registered by its path url, which is very important feature for Ballista to support remote storage, like HDFS, etc. To avoid to introduce the remote object store dependency for DataFusion core, it's better to introduce a trait object for the ObjectStoreRegistry to self detect ObjectStore by url. Then we can introduce the dependencies into the Ballista rather than the DataFusion core with a self-registry implementation.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #2905.

 # Rationale for this change

Currently ObjectStore cannot be self-registered by its path url, which is very important feature for Ballista to support remote storage, like HDFS, etc. To avoid to introduce the remote object store dependency for DataFusion core, it's better to make the *ObjectStoreRegistry* have the ability to self detect an ObjectStore based on url.

# What changes are included in this PR?

Introduce a trait property, *ObjectStoreSelfDetector*, for the *ObjectStoreRegistry* for the ability to self detect an ObjectStore based on url. If Ballista wants to have this ability, it needs to set a specific detector for the *ObjectStoreRegistry* and config it in the *RumtimeEnv*.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by yahoNanJing** [comment]
Hi @alamb, @thinkharderdev, @andygrove, could help review this PR? Thanks in advance.

**Comment by alamb** [comment]
cc @tustvold

File: Cargo.toml [hunk 1]
```diff
@@ -19,7 +19,6 @@
 members = [
     "datafusion/common",
     "datafusion/core",
-    "datafusion/data-access",
```
Comment (alamb) [inline Cargo.toml:22]: this code is unused, correct? And this is a cleanup that is not strictly necessary for this PR? (this is fine, I am just trying to make sure I understand)

**Review by alamb — COMMENTED** [review]
Thanks @yahoNanJing  I started to look at this PR and the code seems to do what is described in the description 👍  but I was not quite sure I understand the usecase. 

Is the primary usecase  "lazy" object store registration? Namely that the relevant `ObjectStore` instance isn't instantiated unless a URL in a plan requires it? 

> To avoid to introduce the remote object store dependency for DataFusion core, 

I agree this is not good. 

I thought the way Ballista (or other systems) would work would be that:

1. The user would provide a configuration file for whatever object stores (e.g. HDFS credentials, or similar) they wanted to connect to
2. On process startup, Ballista would create `ObjectStore` instances based on the configuration and register them with the global `ObjectStoreRegistry`
3. Since all Ballista processes would have the same (or compatible) configuration, serialized urls created by one would have the correct `ObjectStore` already instantiated int hem

File: Cargo.toml (hunk 1 above)
Comment (yahoNanJing) [reply Cargo.toml:22]: Yes.  This PR is based on another PR #2904

**Comment by yahoNanJing** [comment]
Thanks @alamb for your comments.

> Is the primary usecase "lazy" object store registration? Namely that the relevant ObjectStore instance isn't instantiated unless a URL in a plan requires it?

Yes. By self-registration, we can achieve the "lazy" registration.

> I thought the way Ballista (or other systems) would work would be that:

For the ballista usage, it's necessary for us to introduce the object store extensions as optional features. The detailed usage way may differ for different object stores. For example,
- For S3, we need to introduce additional configurations, like bucket, credential info, etc.
- For HDFS, related configurations can be found in some specific locations. Therefore, we don't need to do any additional configurations.

**Review by tustvold — COMMENTED** [review]
If I understand correctly the issue is following switching to `object_store`, the stores have to be registered per-host whereas previously they were registered per-scheme. The desired functionality is to be able to register a generic connector that knows given any hdfs URL how to create an appropriate `ObjectStore` for it.

As such I wonder if we might instead call this concept `ObjectStoreSchemeProvider` and store a HashMap on `ObjectStoreRegistry`? What do you think?

**Comment by alamb** [comment]
> If I understand correctly the issue is following switching to object_store, the stores have to be registered per-host whereas previously they were registered per-scheme. The desired functionality is to be able to register a generic connector that knows given any hdfs URL how to create an appropriate ObjectStore for it.

Is this correct @yahoNanJing ? If so then it makes sense to me ( I didn't realize it was per host rather than per-scheme) Does that means it requires an `ObjectStore` instance per each HDFS URL (`hdfs://10.10.0.1` and `hdfs://10.10.0.2` would be different ObjectStore instances)?

**Comment by tustvold** [comment]
>  per each HDFS URL (hdfs://10.10.0.1 and hdfs://10.10.0.2 would be different ObjectStore instances)?

Correct, we only pass the URL path to `ObjectStore`, and not any authority (host, port, credentials, etc...).

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 2]
```diff
@@ -162,7 +162,7 @@ pub fn split_files(
 pub async fn pruned_partition_list<'a>(
     store: &'a dyn ObjectStore,
     table_path: &'a ListingTableUrl,
-    filters: &[Expr],
+    filters: &'a [Expr],
```
[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/listing/helpers.rs:165]: When compiling, if without this, it throws lifetime exception.

File: datafusion/core/src/datasource/object_store.rs [hunk 3]
```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
...
+    fn get_by_url(&self, url: &Url) -> Option<(String, Arc<dyn ObjectStore>)>;
+}
+
 /// Object store registry
+#[derive(Clone)]
 pub struct ObjectStoreRegistry {
     /// A map from scheme to object store that serve list / read operations for the store
-    object_stores: RwLock<HashMap<String, Arc<dyn ObjectStore>>>,
+    object_stores: Arc<RwLock<HashMap<String, Arc<dyn ObjectStore>>>>,
+    self_detector: Option<Arc<dyn ObjectStoreSelfDetector>>,
```
[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:96]: I think string is OK

File: datafusion/core/src/datasource/object_store.rs [hunk 4]
```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
     }
 }
 
+/// Object store self detector can detector an object store based on the url
+pub trait ObjectStoreSelfDetector: Send + Sync + 'static {
```
[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:85]: Actually the functionality for this trait is for getting object store from url.

**Comment by yahoNanJing** [comment]
> If I understand correctly the issue is following switching to object_store, the stores have to be registered per-host whereas previously they were registered per-scheme. The desired functionality is to be able to register a generic connector that knows given any hdfs URL how to create an appropriate ObjectStore for it.

Hi @alamb and @tustvold, as @tustvold mentioned above, the switching to the object_store changes the previous behavior. And now we need to create ObjectStore for each host.

File: datafusion/core/src/datasource/listing/helpers.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/core/src/datasource/listing/helpers.rs:165]: I think it is a rust complier bug, the original code should not compile.

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/datasource/listing/helpers.rs:165]: I also see this error occasionally (but not always) locally

File: datafusion/core/src/datasource/object_store.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/datasource/object_store.rs:85]: I think @tustvold 's suggestion is to use a slightly different name that perhaps is more conventional (I think often the term "Provider" is used to describe factory-like things in object oriented programming that instantiate instances of interfaces -- aka what this is doing)

File: datafusion/core/src/datasource/object_store.rs [hunk 5]
```diff
@@ -132,19 +149,51 @@ impl ObjectStoreRegistry {
...
-                "No suitable object store found for {}",
-                url
-            ))
-        })?;
-
-        Ok(store.clone())
+        // First check whether can get object store from registry
+        let store = {
+            let stores = self.object_stores.read();
+            let s = &url[url::Position::BeforeScheme..url::Position::BeforeHost];
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/datasource/object_store.rs:159]: What if we called the detector *first* and if that does not return an instance, we can fall back to the `self.object_stores`? This would preserve the existing behavior and would likely have a simpler implementation

Something like (untested): 

```rust
let store = self_detector
  // try to get url from detector first
  .map(|detector| detector.get_by_url(url))
  // fallback to looking up by registered scheme
  .or_else(|| self.oject_stores.read().get(s))
  .ok_or_else(|| {
                DataFusionError::Internal(format!(
                    "No suitable object store found for {}",
                    url
                ))
            })?;
```

**Comment by alamb** [comment]
> As such I wonder if we might instead call this concept ObjectStoreSchemeProvider and store a HashMap on ObjectStoreRegistry? What do people think? This would also allow creating an S3SchemeProvider that knows how to connect to any S3 bucket, etc...

Sounds like a good idea to me -- depending on @yahoNanJing 's preference, we could also add the feature to DataFusion and then consolidate when it was added / integrated into `object_store`

**Review by alamb — COMMENTED** [review]
Thanks @yahoNanJing  -- I understand the need for this PR and would like to get it in to unblock you if needed

The only thing that this PR needs, in my opinion,  is a test (of the ObjectStoreDetector logic) 

 @tustvold  and I had some suggestions on how to improve the code but I think they could be done as a follow on PR as well

File: datafusion/core/src/datasource/object_store.rs [hunk 6]
```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
     }
 }
 
+/// Object store self detector can detector an object store based on the url
```
Comment (alamb) [inline datafusion/core/src/datasource/object_store.rs:84]:
```suggestion
/// Object store self detector can detector an object store based on the url
///
/// This supports the usecase of providing [`ObjectStore`] instances based on  different hosts
//  (such as `hdfs://10.10.0.1` and `hdfs://10.10.0.2` which might have different access access credentials)?

Correct, we only pass the URL path to ObjectStore, and not any authority (host, port, credentials, etc...).
```

**Comment by yahoNanJing** [comment]
> Thanks @yahoNanJing -- I understand the need for this PR and would like to get it in to unblock you if needed
> 
> The only thing that this PR needs, in my opinion, is a test (of the ObjectStoreDetector logic)
> 
> @tustvold and I had some suggestions on how to improve the code but I think they could be done as a follow on PR as well

Actually, this feature is only ballista's interests currently. And I'll add a test in the ballista repo. However, due to the dependency issue among datafusion, ballista and datafusion-objectstore-hdfs. It's necessary for us to merge this PR first. 😢

File: datafusion/core/src/datasource/object_store.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:85]: Actually it's based on the url rather than the scheme. How about just rename it to be ObjectStoreProvider

File: datafusion/core/src/datasource/object_store.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:159]: I think it's better to call the detector later so that we can reuse the registered object stores.

**Comment by yahoNanJing** [comment]
For someone may be interested, a initial implementation with the integration of datafusion, ballista and objectstore-hdfs is in https://github.com/yahoNanJing/arrow-ballista/tree/dev-202207.

File: datafusion/core/src/datasource/object_store.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/datasource/object_store.rs:159]: What I don't like about the current code is that it is hard to concisely explain how urls are resolved as various parts of the URL are tried in a sequence that I find confusing. 

I guess I was thinking that we could keep the logic in DataFusion simple,  Ballista or some other system could implement whatever arbitrary logic was needed in its implementation of the detector. If the detector wanted to first try the registered object stores it could do so.

[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:159]: The previous logic is not correct for the hdfs case, like hdfs://localhost:8020/xxx/xxx. The key needs to include the port info.

[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/core/src/datasource/object_store.rs:159]: Oh sorry, that's my mistake - it should be BeforePath not AfterHost, agreed 👍

[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:159]: If use before path, then it may include credential infos to the key for S3, right? 😭 It's also not good, right?

[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/core/src/datasource/object_store.rs:159]: Why is that not good? I'd expect to be able to register two object stores with the same host but different credentials? I don't follow why this is an issue?

[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:159]: It will easily cause security issues. In general, we should not store credentials in memory, especially directly depending on it.

[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/core/src/datasource/object_store.rs:159]: Could you perhaps expand on the threat model here, and how this would impact it. If the user has opted to use HTTP Basic Auth, I don't see how storing that data in memory as part of the URL is avoidable...

[the remark this answers is not in the document]
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:159]: In general, we should not store the users' credential info in memory. The security topic related to many points:
- The expiration mechanism
- The encoding algorithm
- The encryption algorithm
- ...

However, this PR is not for dealing with it. If anyone is interested in it, we should open another issue. For HDFS, we leverage Kerberos technique to deal with security issue. However, I have little knowledge for S3. My intuition tells me that it's not safe to put the credential info in each request, although it's base64 encoded.

For this PR, it's OK to just use *url::Position::BeforeScheme..url::Position::BeforePath* to extract the key for one specific object store. And I'll submit a related commit to fix it.

File: datafusion/core/src/datasource/object_store.rs [hunk 7]
```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
     }
 }
 
+/// Object store provider can detector an object store based on the url
+pub trait ObjectStoreProvider: Send + Sync + 'static {
+    /// Detector a suitable object store based on its url if possible
+    /// Return the key and object store
+    fn get_by_url(&self, url: &Url) -> Option<(String, Arc<dyn ObjectStore>)>;
```
Comment (tustvold) [inline datafusion/core/src/datasource/object_store.rs:88]: I'm not sure how this returning of a `String` is intended to work, as it will always be looked up in the hashmap based on `&url[url::Position::BeforeScheme..url::Position::BeforePath];`

File: datafusion/core/src/datasource/object_store.rs [hunk 8]
```diff
@@ -105,13 +114,21 @@ impl Default for ObjectStoreRegistry {
...
 impl ObjectStoreRegistry {
+    /// By default the self detector is None
+    pub fn new() -> Self {
+        ObjectStoreRegistry::new_with_detector(None)
+    }
+
     /// Create the registry that object stores can registered into.
     /// ['LocalFileSystem'] store is registered in by default to support read local files natively.
-    pub fn new() -> Self {
+    pub fn new_with_detector(
```
Comment (tustvold) [inline datafusion/core/src/datasource/object_store.rs:124]:
```suggestion
    pub fn new_with_provider(
```

File: datafusion/core/src/datasource/object_store.rs [hunk 9]
```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
...
+    fn get_by_url(&self, url: &Url) -> Option<(String, Arc<dyn ObjectStore>)>;
+}
+
 /// Object store registry
+#[derive(Clone)]
 pub struct ObjectStoreRegistry {
     /// A map from scheme to object store that serve list / read operations for the store
-    object_stores: RwLock<HashMap<String, Arc<dyn ObjectStore>>>,
+    object_stores: Arc<RwLock<HashMap<String, Arc<dyn ObjectStore>>>>,
+    self_detector: Option<Arc<dyn ObjectStoreProvider>>,
```
Comment (tustvold) [inline datafusion/core/src/datasource/object_store.rs:96]:
```suggestion
    provider: Option<Arc<dyn ObjectStoreProvider>>,
```

File: datafusion/core/src/datasource/object_store.rs [hunk 10]
```diff
@@ -132,19 +149,41 @@ impl ObjectStoreRegistry {
...
-        Ok(store.clone())
+        // First check whether can get object store from registry
+        let store = {
+            let stores = self.object_stores.read();
+            let s = &url[url::Position::BeforeScheme..url::Position::BeforePath];
+            stores.get(s).cloned()
+        };
+
+        // If not, then try to detector based on its url.
+        let store = store
```
Comment (tustvold) [inline datafusion/core/src/datasource/object_store.rs:164]: This currently has a thread race, it needs to:

* Acquire the write lock
* Verify that a value hasn't been inserted in the intervening time
* Call the detector
* Add the result
* Drop the write lock

To be honest, this could probably switch to just using a `Mutex` to keep things simple. There is unlikely to be significant contention to warrant the RWLock

File: datafusion/core/src/datasource/object_store.rs [hunk 11]
```diff
@@ -81,10 +81,19 @@ impl std::fmt::Display for ObjectStoreUrl {
...
 
+/// Object store provider can detector an object store based on the url
+pub trait ObjectStoreProvider: Send + Sync + 'static {
+    /// Detector a suitable object store based on its url if possible
+    /// Return the key and object store
+    fn get_by_url(&self, url: &Url) -> Option<(String, Arc<dyn ObjectStore>)>;
+}
+
 /// Object store registry
+#[derive(Clone)]
```
Comment (tustvold) [inline datafusion/core/src/datasource/object_store.rs:92]: Why is this needed, afaict all locations use `Arc<ObjectStoreRegistry>`?

File: datafusion/core/src/execution/runtime_env.rs [hunk 12]
```diff
@@ -121,6 +122,8 @@ pub struct RuntimeConfig {
     pub disk_manager: DiskManagerConfig,
     /// MemoryManager to limit access to memory
     pub memory_manager: MemoryManagerConfig,
+    /// ObjectStoreRegistry to get object store based on url
+    pub object_store_registry: ObjectStoreRegistry,
```
Comment (tustvold) [inline datafusion/core/src/execution/runtime_env.rs:126]: This should possibly be `Arc<ObjectStoreRegistry>` for consistency

File: datafusion/core/src/datasource/object_store.rs (hunk 7 above)
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:88]: You are right. With the new logic of get_by_url, we can just use the key as the parameter and the returned value without the String key.

File: datafusion/core/src/datasource/object_store.rs (hunk 10 above)
Comment (yahoNanJing) [reply datafusion/core/src/datasource/object_store.rs:164]: I don't think so. The RWLock is better than just using Mutex. Here, the read case will happen frequently. While the write case only happens a few times.

File: datafusion/core/src/execution/runtime_env.rs (hunk 12 above)
Comment (yahoNanJing) [reply datafusion/core/src/execution/runtime_env.rs:126]: Why? Other properties are not wrapped with Arc. And for one env, there should be only one runtime_env and its related properties.

File: datafusion/core/src/datasource/object_store.rs (hunk 10 above)
Comment (tustvold) [reply datafusion/core/src/datasource/object_store.rs:162]: In the uncontended case, which is extremely likely given how short the critical section is, they will perform exactly the same - if anything the Mutex might be marginally faster. It was more an observation that the complexity of using a RWLock is probably not actually yielding any return

File: datafusion/core/src/execution/runtime_env.rs (hunk 12 above)
Comment (tustvold) [reply datafusion/core/src/execution/runtime_env.rs:126]: Two reasons:

* We use `Arc<ObjectStoreRegistry>` elsewhere and so using it here is more consistent
* It provides a hint that this is shared state, e.g. we use `Arc<DiskManager>` instead of just `DiskManager`, etc...

File: datafusion/core/src/catalog/schema.rs [hunk 13]
```diff
@@ -162,7 +162,17 @@ mod tests {
     #[tokio::test]
     async fn test_schema_register_listing_table() {
         let testdata = crate::test_util::parquet_test_data();
-        let filename = format!("file:///{}/{}", testdata, "alltypes_plain.parquet");
+        let testdir = if testdata.starts_with('/') {
```
Comment (tustvold) [inline datafusion/core/src/catalog/schema.rs:165]: I wonder if this logic would be better handled in get_data_dir (to remove any trailing /)? As it stands I'm surprised just making this change here is sufficient?

**Review by tustvold — APPROVED** [review]
I think lets get this in and we can refine it in follow up PRs

File: datafusion/core/src/execution/runtime_env.rs (hunk 12 above)
Comment (yahoNanJing) [reply datafusion/core/src/execution/runtime_env.rs:126]: I see. Agree with you that here better to use Arc

File: datafusion/core/src/catalog/schema.rs (hunk 13 above)
Comment (yahoNanJing) [reply datafusion/core/src/catalog/schema.rs:165]: For different OS, the detailed parsing logic is different. It seems the logic here is too trivial and should be refined.
